rules/numeric-separators-style.js

```diff
@@ -0,0 +1,142 @@
+'use strict';
+const merge = require('lodash/merge');
+const getDocumentationUrl = require('./utils/get-documentation-url');
+
+function addSeparator(value, {minimumDigits, groupLength}, fromLeft) {
+	const {length} = value;
+
+	if (length < minimumDigits) {
+		return value;
+	}
+
+	const parts = [];
+	if (fromLeft) {
+		for (let start = 0; start < length; start += groupLength) {
+			const end = Math.min(start + groupLength, length);
+			parts.push(value.slice(start, end));
+		}
+	} else {
+		for (let end = length; end > 0; end -= groupLength) {
+			const start = Math.max(end - groupLength, 0);
+			parts.unshift(value.slice(start, end));
+		}
+	}
+
+	return parts.join('_');
+}
+
+function addSeparatorFromLeft(value, options) {
+	return addSeparator(value, options, true);
+}
+
+function formatNumber(value, options) {
+	const parts = value.split('.');
+	const [decimal, fractional] = parts;
+
+	let formatted = addSeparator(decimal, options);
+	if (parts.length === 2) {
+		formatted += '.';
+		formatted += addSeparatorFromLeft(fractional, options);
+	}
+
+	return formatted;
+}
+
+function format(value, options) {
+	const {
+		prefix = '',
+		data
+	} = value.match(/^(?<prefix>0[box])?(?<data>.*)$/i).groups;
+
+	const formatOption = options[prefix.toLowerCase()];
+
+	if (prefix) {
+		return prefix + addSeparator(data, formatOption);
+	}
+
+	const {
+		number,
+		mark = '',
+		sign = '',
+		power = ''
+	} = value.match(/^(?<number>.*?)(?:(?<mark>e)(?<sign>[+-]?)?(?<power>\d+))?$/i).groups;
+
+	return formatNumber(number, formatOption) + mark + sign + addSeparator(power, options['']);
+}
+
+const create = context => {
+	const defaultOptions = {
+		hexadecimal: {minimumDigits: 0, groupLength: 2},
+		binary: {minimumDigits: 0, groupLength: 4},
+		octal: {minimumDigits: 0, groupLength: 4},
+		number: {minimumDigits: 5, groupLength: 3}
+	};
+	const rawOptions = merge(defaultOptions, context.options[0]);
```

Comment:
`_.merge` and `_.defaultsDeep` which is better in this case?  Not sure about difference between them.

Reply:
Looking at lodash's code, merge might be slightly faster, because `_.defaultsDeep` is base on `_.mergeWith`, which is based on the exact same function as `_.merge`.

Reply:
We  need deep merge, right? Does `_.merge` do that?

Reply:
And make sure it don't modify defaultOptions

Reply:
Seen  this?  Is it the reason for  https://github.com/sindresorhus/eslint-plugin-unicorn/pull/833#discussion_r495986350?

Reply:
Oh I missed your messages! Sorry. It might, let me try, I'll get back to you in a minute

Reply:
Yes, `_.merge` does deep merge.
`_.defaultsDeep` seems to have a little problem, if there are multiple properties, and won't modify them all:
If i have `context.options[0] { number: { minimumDigits: 6 } }`, it will give me `rawOptions { number: { minimumDigits: 6 } }`, which is not what we want.
I have 3 failure in the tests with `_.defaultsDeep`.